File: example/xx_three/two/performance.js
// performance.js

var viewer = two.ui.scene_viewer(app.ui.begin());
//two.ui.orbit_controller(viewer);

var scene = viewer.scene;

//scene.background = new THREE.Color(0xffffff);

var objects = [];

if (init) {
    this.mouse = new two.vec2(0.0);

    var camera = viewer.camera;
    camera.fov = 60.0; camera.near = 1.0; camera.far = 10000.0;
    camera.eye.z = 3200.0;

    var normal = app.gfx.programs.fetch('normal');
    //var normal = app.gfx.programs.fetch('pbr/pbr');

    var material = app.gfx.materials.create('normal');
    var m = material; m.program = normal;

    var suzanne = two.gfx.model_suzanne(app.gfx);

    for(var i = 0; i < 5000; i++)
    {
        var p = new two.vec3(Math.random() * 8000.0 - 4000.0, Math.random() * 8000.0 - 4000.0, Math.random() * 8000.0 - 4000.0);
        var a = new two.vec3(Math.random() * 2 * Math.PI, Math.random() * 2 * Math.PI, 0.0);
        var s = new two.vec3(Math.random() * 50 + 100);

        var n = scene.nodes().add(new two.Node3(p, new two.quat(a), s));
        var it = scene.items().add(new two.Item(n, suzanne, 0, material));

        objects.push({ node: n, position: p, angles: a, scale: s });
    }
}

var event = viewer.mouse_event(two.DeviceType.Mouse, two.EventType.Moved);
if(event.valid())
{
    this.mouse.x = (event.relative.x - viewer.frame.size.x / 2.0) * 10.0;
    this.mouse.y = (event.relative.y - viewer.frame.size.y / 2.0) * 10.0;
}

var camera = viewer.camera;
camera.eye.x += (mouse.x - camera.eye.x) * .05;
camera.eye.y += (-mouse.y - camera.eye.y) * .05;

for(var i = 0; i < objects.length; ++i)
{
    var o = objects[i];
    o.angles.x += 0.01;
    o.angles.y += 0.02;

    o.node.apply(o.position, new two.quat(o.angles), o.scale);
}

var root = scene.begin();
two.gfx.radiance(root, 'radiance/tiber_1_1k.hdr', two.BackgroundMode.Radiance);

//two.gfx.shape(root, new two.Sphere(), new two.Symbol(), 0, material);

File: example/xx_three/performance.cpp
//#include <mud/frame.h>
#include <frame/Api.h>
#include <gfx-pbr/Api.h>

#include <xx_three/xx_three.h>

#include <stl/array.h>
#include <stl/vector.hpp>

using namespace mud;

void xx_performance(Shell& app, Widget& parent, Dockbar& dockbar)
{
	UNUSED(dockbar);
	SceneViewer& viewer = ui::scene_viewer(parent);
	//ui::orbit_controller(viewer);

	Scene& scene = viewer.m_scene;

	//scene.background = new THREE.Color(0xffffff);

	struct Object { Node3* node; vec3 position; vec3 angles; vec3 scale; };
	static vector<Object> objects = {};

	static Program& normal = app.m_gfx.programs().fetch("normal");

	static Material& material = app.m_gfx.materials().create("normal", [&](Material& m) {
		m.m_program = &normal;
	});

	static Model& suzanne = gfx::model_suzanne(app.m_gfx);

	static bool once = false;
	if(!once)
	{
		once = true;

		Camera& camera = viewer.m_camera;
		camera.m_near = 1.f; camera.m_far = 10'000.f; camera.m_fov = 60.f;
		camera.m_eye.z = 3'200.f;

		for(size_t i = 0; i < 5000; i++)
		{
			vec3 position = vec3(randf(), randf(), randf()) * 8000.f - 4000.f;
			vec3 angles = vec3(randf() * c_2pi, randf() * c_2pi, 0.f);
			vec3 scale = vec3(randf() * 50 + 100);

			Node3& n = gfx::nodes(scene).add(Node3(position, quat(angles), scale));
			Item& it = gfx::items(scene).add(Item(n, suzanne, 0U, &material));

			objects.push_back({ &n, position, angles, scale });
		}
	}

	static vec2 mouse = vec2(0.f);
	if(MouseEvent event = viewer.mouse_event(DeviceType::Mouse, EventType::Moved))
	{
		mouse = (event.m_relative - viewer.m_frame.m_size / 2.f) * 10.f;
	}

	Camera& camera = viewer.m_camera;
	camera.m_eye.x += (mouse.x - camera.m_eye.x) * .05f;
	camera.m_eye.y += (-mouse.y - camera.m_eye.y) * .05f;
	camera.m_target = vec3(0.f); // scene.position;

	for(Object& o : objects)
	{
		o.angles.x += 0.01;
		o.angles.y += 0.02;

		o.node->m_transform = bxTRS(o.scale, quat(o.angles), o.position);
	}

	Gnode& root = scene.begin();
	gfx::radiance(root, "radiance/tiber_1_1k.hdr", BackgroundMode::Radiance);

	gfx::shape(root, Sphere(), Symbol(), 0U, &material);
}